packages/core/src/tracing/instrumentation/database/prisma.js

```diff
@@ -0,0 +1,204 @@
+          );
+        }
+      }
+    }
+    prismaClientModule.PrismaClient = InstanaPrismaClient;
+  }
+}
+
+function shimRequest(prismaClientModule) {
+  if (
+    !prismaClientModule ||
+    !prismaClientModule.PrismaClient ||
+    !prismaClientModule.PrismaClient.prototype ||
+    !prismaClientModule.PrismaClient.prototype._request
+  ) {
+    logger.debug('prismaClientModule.PrismaClient.prototype._request does not exist, will not instrument Prisma.');
+    return;
+  }
+
+  // The relevant source file is:
+  // https://github.com/prisma/prisma/blob/main/packages/client/src/runtime/getPrismaClient.ts
+  shimmer.wrap(prismaClientModule.PrismaClient.prototype, '_request', instrumentRequest);
+}
+
+function instrumentRequest(original) {
+  return function () {
+    if (cls.skipExitTracing({ isActive })) {
+      return original.apply(this, arguments);
+    }
+
+    const argsForOriginalRequest = new Array(arguments.length);
+    for (let i = 0; i < arguments.length; i++) {
+      argsForOriginalRequest[i] = arguments[i];
+    }
+    return instrumentedRequest(this, original, argsForOriginalRequest);
+  };
+}
+
+function instrumentedRequest(ctx, originalRequest, argsForOriginalRequest) {
+  return cls.ns.runAndReturn(() => {
+    const span = cls.startSpan('prisma', EXIT);
+    span.stack = getStackTrace(instrumentedRequest, 1);
+    const params = argsForOriginalRequest[0] || {};
+    span.data.prisma = {
+      model: params.model,
+      action: params.action,
+      provider: ctx.__inProvider,
+      url: ctx.in_DataSourceUrl
+    };
+    const requestPromise = originalRequest.apply(ctx, argsForOriginalRequest);
+    if (!requestPromise && typeof requestPromise.then !== 'function') {
+      span.cancel();
+      return requestPromise;
+    } else {
+      return requestPromise
+        .then(value => {
+          finishSpan(null, span);
+          return value;
+        })
+        .catch(error => {
+          finishSpan(error, span);
+          return error;
+        });
+    }
+  });
+}
+
+function redactPassword(provider, url) {
+  if (typeof provider !== 'string' || typeof url !== 'string' || 'sqlite' === provider.toLowerCase()) {
+    return url;
+  }
+
+  const isKnownProvider = ['cockroachdb', 'mongodb', 'mysql', 'postgresql', 'sqlserver'].includes(
+    provider.toLowerCase()
+  );
+
+  if (!isKnownProvider) {
+    // Prisma might add new providers in the future and we cannot know how the connection URLs are structured. To err on
+    // the side of caution, we do not capture the datasource URL for unknown providers. We do not want to risk capturing
+    // the DB password inadvertently.
+    return null;
+  }
+
+  if ('sqlserver' === provider.toLowerCase()) {
+    return redactPasswordFromMsSQLUrl(url);
+  }
+
+  // Parse standard datasource URI (basically everything except for MsSQL).
+  try {
+    const parsedUrl = new URL(url);
+
+    if (parsedUrl != null) {
+      parsedUrl.password = '_redacted_';
```

Comment:
Right, I went with `<redacted>` first but I ran into some issues because `<` and `>` are not allowed in URLs. That's why I decided to go with the safer variant `_redacted_`. We parse the URI in the back end to extract host and port for infrastructure correlation, and that will only work with a valid URI, using `<redacted>` would break that. Alternatively, we could percent-encode `<redacted>` but I think `_redacted_` is the better choice over `%3Credacted%3E`.

Reply:
I'll add a comment about that.